test/gurka/test_reproduce_tile_build.cc

```diff
@@ -0,0 +1,99 @@
+
+#if !defined(VALHALLA_SOURCE_DIR)
+#define VALHALLA_SOURCE_DIR
+#endif
+
+using namespace valhalla;
+using namespace valhalla::baldr;
+
+void assert_tile_equalish(const GraphTile& a, const GraphTile& b) {
+  // expected size
+  ASSERT_EQ(a.header()->end_offset(), b.header()->end_offset());
+
+  // check the first chunk after the header
+  ASSERT_EQ(memcmp(reinterpret_cast<const char*>(a.header()) + sizeof(GraphTileHeader),
+                   reinterpret_cast<const char*>(b.header()) + sizeof(GraphTileHeader),
+                   (reinterpret_cast<const char*>(b.GetBin(0, 0).begin()) -
+                    reinterpret_cast<const char*>(b.header())) -
+                       sizeof(GraphTileHeader)),
+            0);
+
+  // check the stuff after the bins
+  ASSERT_EQ(memcmp(reinterpret_cast<const char*>(a.header()) + a.header()->edgeinfo_offset(),
+                   reinterpret_cast<const char*>(b.header()) + b.header()->edgeinfo_offset(),
+                   b.header()->end_offset() - b.header()->edgeinfo_offset()),
+            0);
+
+  // if the header is as expected
+  const auto *ah = a.header(), *bh = b.header();
+  if (ah->access_restriction_count() == bh->access_restriction_count() &&
+      ah->admincount() == bh->admincount() &&
+      ah->complex_restriction_forward_offset() == bh->complex_restriction_forward_offset() &&
+      ah->complex_restriction_reverse_offset() == bh->complex_restriction_reverse_offset() &&
+      ah->date_created() == bh->date_created() && ah->density() == bh->density() &&
+      ah->departurecount() == bh->departurecount() &&
+      ah->directededgecount() == bh->directededgecount() &&
+      ah->edgeinfo_offset() == bh->edgeinfo_offset() && ah->exit_quality() == bh->exit_quality() &&
+      ah->graphid() == bh->graphid() && ah->name_quality() == bh->name_quality() &&
+      ah->nodecount() == bh->nodecount() && ah->routecount() == bh->routecount() &&
+      ah->signcount() == bh->signcount() && ah->speed_quality() == bh->speed_quality() &&
+      ah->stopcount() == bh->stopcount() && ah->textlist_offset() == bh->textlist_offset() &&
+      ah->schedulecount() == bh->schedulecount() && ah->version() == bh->version()) {
+    // make sure the edges' shape and names match
+    for (size_t i = 0; i < ah->directededgecount(); ++i) {
+      auto a_info = a.edgeinfo(a.directededge(i)->edgeinfo_offset());
+      auto b_info = b.edgeinfo(b.directededge(i)->edgeinfo_offset());
+      ASSERT_EQ(a_info.encoded_shape(), b_info.encoded_shape());
+      ASSERT_EQ(a_info.GetNames().size(), b_info.GetNames().size());
+      for (size_t j = 0; j < a_info.GetNames().size(); ++j)
+        ASSERT_EQ(a_info.GetNames()[j], b_info.GetNames()[j]);
+    }
+  } else {
+    FAIL() << "not equal";
+  }
+}
+
+// 1. build tiles with the same input twice
+// 2. check that the same tile sets are generated
+TEST(GraphTileBuilder, TestBuildIsReproducible) {
+  const auto build_tiles = [](const std::string& dir) -> gurka::map {
```

Comment:
There's a lot of `auto` here! There are mixed feelings about this amongst the contributors: some really like `auto` everywhere and others think it should be confined to variable-capture expressions (like `for` loops, etc.), STL-iterator return types, etc. I'm in the second category, and would prefer to use explicit types here when possible, because it'll make the code easier to read by people unfamiliar with the types in the program.

Reply:
Sure, done. I left 2 lambdas declared with auto(if you don't mind), all other types(including valhalla types are explicit now)

Reply:
Ah yeah, for the lambda, totally makes sense.